openedx/core/djangoapps/course_date_signals/tests.py

```diff
@@ -114,5 +116,261 @@ def test_sequence_with_graded_and_ungraded_assignments(self):
+                ItemFactory.create(category='problem', parent=vertical1)
+
+                chapter2 = ItemFactory.create(category='chapter', parent=course10)
+                sequential2 = ItemFactory.create(category='sequential', parent=chapter2, graded=True)
+                vertical2 = ItemFactory.create(category='vertical', parent=sequential2)
+                ItemFactory.create(category='problem', parent=vertical2)
+
+                chapter3 = ItemFactory.create(category='chapter', parent=course10)
+                sequential3 = ItemFactory.create(category='sequential', parent=chapter3, graded=True)
+                vertical3 = ItemFactory.create(category='vertical', parent=sequential3)
+                ItemFactory.create(category='problem', parent=vertical3)
+
+                chapter4 = ItemFactory.create(category='chapter', parent=course10)
+                sequential4 = ItemFactory.create(category='sequential', parent=chapter4, graded=True)
+                vertical4 = ItemFactory.create(category='vertical', parent=sequential4)
+                ItemFactory.create(category='problem', parent=vertical4)
+            self.courses.append(course10)
+
+
+    @override_waffle_flag(CUSTOM_PLS, active=True)
+    def test_extract_dates_from_course(self):
+        """
+        extract_dates_from_course should return a list of (block item location, field metadata dictionary)
+        """
+
+        # course 1: With due_num_weeks but without any units
+        course = self.courses[0]
+        chapter = course.get_children()[0]
+        sequential = chapter.get_children()[0]
+        expected_dates = [
+            (course.location, {}),
+            (chapter.location, timedelta(days=28)),
+            (sequential.location, {'due': timedelta(days=21)})
+        ]
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+        # with due_num_weeks and a unit
+        vertical = ItemFactory.create(category='vertical', parent=sequential)
+        expected_dates.append((vertical.location, {'due': timedelta(days=21)}))
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+        # with due_num_weeks and a problem
+        problem = ItemFactory.create(category='problem', parent=vertical)
+        expected_dates.append((problem.location, {'due': timedelta(days=21)}))
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
+
+        # due_num_weeks in one of the sections
+        sequential2 = ItemFactory.create(category='sequential', parent=chapter)
+        sequential3 = ItemFactory.create(category='sequential', parent=chapter)
+        course = modulestore().get_item(course.location)
+        self.assertCountEqual(extract_dates_from_course(course), expected_dates)
```

Comment:
Wait, why don't these subsections get due dates?

Reply:
I was testing when only one of the subsections has due dates (so only `sequential1` has a due date and `sequential2` and `sequential3` don't). And since `sequential2` and `sequential3` are not graded the current PLS wouldn't apply either.

Reply:
Oh my explanation comments of what I am testing is confusing! Will fix that